install.sh doesn't remove the old igb_uio module and install the new one when compiling dpdk

## Bug Report

**Current Behavior**
`install.sh` doesn't reinstall igb_uio module generated by dpdk if it finds the system has already installed one. Since dpdk provides the driver of igb_uio which might be often updated, the old igb_uio module should be removed and replaced by the dpdk's. This can cause a run-time error for running `dpdk_iface_main` to create a dpdk-iface device.

**Expected behavior/code**
igb_uio module should be removed and re-installed at each time of compiling dpdk

**Steps to reproduce**

- Apply for a new machine or reboot a machine from cloudlab, the system has already installed a default igb_uio, use `grep -m 1 "igb_uio" /proc/modules` to display:        

``` igb_uio 2446317 0 - Live 0x0000000000000000 (OX) ```

- Then run `install.sh` to compile dpdk, and use `grep -m 1 "igb_uio" /proc/modules` to display again:     

``` igb_uio 2446317 0 - Live 0x0000000000000000 (OX) ```

The old igb_uio module was not replaced by the new one.

-  Run ` setup_mtcp_onvm_env.sh ` to create a dpdk-iface device, an error happened:
```
xiaosuGW@node2:~/mtcp$ ./setup_mtcp_onvm_env.sh
Checking ldflags.txt...
RTE_SDK env variable is set to /users/xiaosuGW/openNetVM-dev/dpdk
RTE_TARGET env variable is set to x86_64-native-linuxapp-gcc
Are you using an Intel NIC (y/n)? y
Creating dpdk interface entries
make -C /lib/modules/3.13.0-117-generic/build/ M=/users/xiaosuGW/mtcp/dpdk-iface-kmod modules
make[1]: Entering directory `/usr/src/linux-headers-3.13.0-117-generic'
  Building modules, stage 2.
  MODPOST 1 modules
make[1]: Leaving directory `/usr/src/linux-headers-3.13.0-117-generic'
make -C /lib/modules/3.13.0-117-generic/build/ M=/users/xiaosuGW/mtcp/dpdk-iface-kmod modules
make[1]: Entering directory `/usr/src/linux-headers-3.13.0-117-generic'
  Building modules, stage 2.
  MODPOST 1 modules
make[1]: Leaving directory `/usr/src/linux-headers-3.13.0-117-generic'
sudo ./dpdk_iface_main
Removing existing device node entry... not present.
Creating device node entry... done.
Setting permissions on the device node entry... done.
Scanning the system for dpdk-compatible devices... done.
Clearing previous entries
ioctl call failed!
make: *** [run] Error 1

```

**Environment**
<!--- Optional: give us some details about the setup -->
- OS:  cloudlab machines
- onvm version:  develop branch with commit `e305489b76c868e722470c7799625de90d5f37bb`
- dpdk version:   18.11.0
- mtcp:  branch `onvm_build_readme_fix` 

**Possible Solution**
Removing the old igb_uio and re-installing the new one under `$RTE_SDK/$RTE_TARGET/kmod` can resolve this problem.

--- a/scripts/dpdk_helper_scripts.sh
+++ b/scripts/dpdk_helper_scripts.sh
@@ -1,0 +1,96 @@
+#!/bin/bash
+
+HUGEPGSZ=`cat /proc/meminfo  | grep Hugepagesize | cut -d : -f 2 | tr -d ' '`
+
+#
+# Unloads igb_uio.ko.
+#
+remove_igb_uio_module()
+{
+    echo "Unloading any existing DPDK UIO module"
+    /sbin/lsmod | grep -s igb_uio > /dev/null
+    if [ $? -eq 0 ] ; then
+        sudo /sbin/rmmod igb_uio
+    fi
+}
+
+#
+# Creates hugepage filesystem.
+#
+create_mnt_huge()
+{
+    echo "Creating /mnt/huge and mounting as hugetlbfs"
+    sudo mkdir -p /mnt/huge
+
+    grep -s '/mnt/huge' /proc/mounts > /dev/null
+    if [ $? -ne 0 ] ; then
+        sudo mount -t hugetlbfs nodev /mnt/huge
+    fi
+}
+
+#
+# Removes hugepage filesystem.
+#
+remove_mnt_huge()
+{
+    echo "Unmounting /mnt/huge and removing directory"
+    grep -s '/mnt/huge' /proc/mountsi > /dev/null
+    if [ $? -eq 0 ] ; then
+        sudo umount /mnt/huge
+    fi
+
+    if [ -d /mnt/huge ] ; then
+        sudo rm -R /mnt/huge
+    fi
+}
+
+#
+# Removes all reserved hugepages.
+#
+clear_huge_pages()
+{
+    echo > .echo_tmp
+    for d in /sys/devices/system/node/node? ; do
+        echo "echo 0 > $d/hugepages/hugepages-${HUGEPGSZ}/nr_hugepages" >> .echo_tmp
+    done
+    echo "Removing currently reserved hugepages"
+    sudo sh .echo_tmp
+    rm -f .echo_tmp
+
+    remove_mnt_huge
+}
+
+#
+# Creates hugepages on specific NUMA nodes.
+#
+set_numa_pages()
+{
+    echo "Setting up hugepages"
+    sleep 1
+    set +e
+
+    clear_huge_pages
+
+    hp_count="${ONVM_NUM_HUGEPAGES:-1024}"
+
+    echo > .echo_tmp
+    for d in /sys/devices/system/node/node? ; do
+        node=$(basename $d)
+        echo "echo $hp_count > $d/hugepages/hugepages-${HUGEPGSZ}/nr_hugepages" >> .echo_tmp
+    done
+    echo "Reserving hugepages"
+    sudo sh .echo_tmp
+    rm -f .echo_tmp
+
+    create_mnt_huge
+
+    hp_free=$(cat /proc/meminfo | grep HugePages_Free | awk '{print $2}')
+    if [ $hp_free == "0" ]; then
+        echo "Coudn't set up huge pages. Did you try turning it off and on again?"
+        exit 1
+    else
+        echo "Huge pages successfully configured"
+    fi
+
+    set -e
+}

--- a/scripts/dpdk_helper_scripts.sh
+++ b/scripts/dpdk_helper_scripts.sh
@@ -34,7 +34,7 @@ create_mnt_huge()
 remove_mnt_huge()
 {
     echo "Unmounting /mnt/huge and removing directory"
-    grep -s '/mnt/huge' /proc/mountsi > /dev/null
+    grep -s '/mnt/huge' /proc/mounts > /dev/null
     if [ $? -eq 0 ] ; then
         sudo umount /mnt/huge
     fi

--- a/scripts/install.sh
+++ b/scripts/install.sh
@@ -77,6 +77,18 @@ fi
 # Set state variables
 start_dir=$(pwd)
 
+if [ -z "$ONVM_HOME" ]; then
+    echo "Please export \$ONVM_HOME and set it to $start_dir"
+    exit 1
+fi
+
+# Source DPDK helper functions
+. $ONVM_HOME/scripts/dpdk_helper_scripts.sh
+
+set +e
+remove_igb_uio_module
+set -e
+
 # Compile dpdk
 cd $RTE_SDK
 echo "Compiling and installing dpdk in $RTE_SDK"
@@ -99,13 +111,9 @@ sudo -v
 
 cd $start_dir
 
-# Configure HugePages only if user wants to
+# Setup/Check for free HugePages if the user wants to
 if [ -z "$ONVM_SKIP_HUGEPAGES" ]; then
-	hp_size=$(cat /proc/meminfo | grep Hugepagesize | awk '{print $2}')
-	hp_count="${ONVM_NUM_HUGEPAGES:-1024}"
-	echo "Configuring $hp_count hugepages with size $hp_size"
-	sleep 1
-	sudo mkdir -p /mnt/huge
+    set_numa_pages
 fi
 
 grep -m 1 "huge" /etc/fstab | cat
@@ -114,16 +122,6 @@ if [ ${PIPESTATUS[0]} != 0 ] && [ -z "$ONVM_SKIP_FSTAB" ]; then
     echo "Adding huge fs to /etc/fstab"
     sleep 1
     sudo sh -c "echo \"huge /mnt/huge hugetlbfs defaults 0 0\" >> /etc/fstab"
-fi
-
-#Only mount hugepages if user wants to 
-if [ -z "$ONVM_SKIP_HUGEPAGES" ]; then
-	echo "Mounting hugepages"
-	sleep 1
-	sudo mount -t hugetlbfs nodev /mnt/huge
-	echo "Creating $hp_count hugepages"
-	sleep 1
-	sudo sh -c "echo $hp_count > /sys/devices/system/node/node0/hugepages/hugepages-${hp_size}kB/nr_hugepages"
 fi
 
 # Configure local environment

--- a/scripts/setup_environment.sh
+++ b/scripts/setup_environment.sh
@@ -73,20 +73,15 @@ if [ -z "$ONVM_HOME" ]; then
     exit 1
 fi
 
+# Source DPDK helper functions
+. $ONVM_HOME/scripts/dpdk_helper_scripts.sh
+
 # Disable ASLR
 sudo sh -c "echo 0 > /proc/sys/kernel/randomize_va_space"
 
 # Setup/Check for free HugePages if the user wants to
 if [ -z "$ONVM_SKIP_HUGEPAGES" ]; then
-	hp_size=$(cat /proc/meminfo | grep Hugepagesize | awk '{print $2}')
-	hp_count="${ONVM_NUM_HUGEPAGES:-1024}"
-
-	sudo sh -c "echo $hp_count > /sys/devices/system/node/node0/hugepages/hugepages-${hp_size}kB/nr_hugepages"
-	hp_free=$(cat /proc/meminfo | grep HugePages_Free | awk '{print $2}')
-	if [ $hp_free == "0" ]; then
-    	echo "No free huge pages. Did you try turning it off and on again?"
-    	exit 1
-	fi
+    set_numa_pages $hp_count
 fi
 
 # Verify sudo access

